Enhancement: extending `pyevtk` to allow for usage of Parallel VTK files

anlavandier opened the pull request:
Hello, I have been using `pyevtk` to write VTK files without having the VTK library as a dependency and I have been very happy with it. The only thing it was missing in my use case was the ability to write "parallel VTK files" as described on the [Kitware website](https://kitware.github.io/vtk-examples/site/VTKFileFormats/#parallel-file-formats).

This pull request consists of :
* A simple reformat of `examples/image.py`, `pyevtk/__init__.py`, `pyevtk/_version.py`, `setup.py`, `test/dummy.py`, `versionner.py` and `pyevtk/evtk.py`.
* The addition of two classes `VtkParallelFileType` and `VtkParallel` which are heavily borrowed from their serial counterpart but changed to be taylored for parallel VTK files.
* Turning start into a keyword argument in some high-level functions
* Adding a new high-level function to write a parallelVTK file for one of the two grids format.

Commit:
Adding VtkParallelFiles class, high level functions and examples with StructuredGrid and RectilinearGrid.

Commit:
Fix docstring + add back line breaks

renefritze commented:
Thank you for this PR. To make the review process easier, could you please either split the formatting into a separate PR or drop it altogether?

anlavandier commented:
> Thank you for this PR. To make the review process easier, could you please either split the formatting into a separate PR or drop it altogether?

Dropped the reformatting outside of the files I also changed.

renefritze commented on pyevtk/hl.py:
```diff
@@ -46,7 +51,9 @@ def _addDataToFile(vtkFile, cellData, pointData, fieldData=None):
         keys = list(pointData.keys())
         # find first scalar and vector data key to set it as attribute
-        scalars = next((key for key in keys if isinstance(pointData[key], np.ndarray)), None)
+        scalars = next(
```
This is still just a formatting change, right?

anlavandier replied:
It is.

renefritze commented on pyevtk/hl.py:
```diff
@@ -103,6 +139,7 @@ def _appendDataToFile(vtkFile, cellData, pointData, fieldData=None):
 # =================================
 def imageToVTK(
     path,
+    start=(0, 0, 0),
```
Would it be a good idea to put this new arg at the end? Otherwise we might break older, non-keyword usages of this code?

anlavandier replied:
I think that it could be moved after `origin` and `spacing` but I don't see a reason to push it past the `Data` keyword arguments.

anlavandier replied:
I completely understand the concern but it feels weird to me to have a "metadata argument" after "data" arguments. I feel like most people use the keywords for `pointData`, `cellData` and fieldData` because of their similar nature.

renefritze replied:
I could see a compromise where we add a new top level function with the new signature and deprecate the current one. I really do not want to chance scripts breaking everywhere for this package that is mostly in maintenance mode.

anlavandier replied:
As I'm probably one of the only person that are going to use this new feature, maybe the sensible thing is to follow your original idea and simply put the `start` keyword argument at the end. Adding a new high-level function seems like to burdensome especially when it is not clear when or if the deprecated function is going to be removed.

anlavandier replied:
I followed through and put the `start` keyword argument at the end. I purposefully did not change the docstrings because I thing that it makes more sense that way.

renefritze commented on pyevtk/hl.py:
```diff
@@ -184,9 +224,11 @@ def imageToVTK(
 # ==============================================================================
-def gridToVTK(path, x, y, z, cellData=None, pointData=None, fieldData=None):
+def gridToVTK(
+    path, x, y, z, start=(0, 0, 0), cellData=None, pointData=None, fieldData=None
```
Same as `imageToVTK`

anlavandier replied:
I don't think that moving this arguments after `fieldData` makes much sense.

renefritze commented on pyevtk/hl.py:
```diff
@@ -198,6 +240,10 @@ def gridToVTK(path, x, y, z, cellData=None, pointData=None, fieldData=None):
         z coordinate axis.
+    start : tuple, optional
+        start of the coordinates.
+        Used in the distributed conwhere each process
```
`conwhere` seems to be a typo?

anlavandier replied:
Yes, typo is fixed in 80749e7.

renefritze commented on pyevtk/hl.py:
```diff
@@ -415,18 +551,14 @@ def linesToVTK(path, x, y, z, cellData=None, pointData=None, fieldData=None):
-    _addDataToFile(
-        w, cellData=cellData, pointData=pointData, fieldData=fieldData
-    )
+    _addDataToFile(w, cellData=cellData, pointData=pointData, fieldData=fieldData)
```
Only formatting again?

anlavandier replied:
Absolutely.

renefritze reviewed (commented):
I just have some minor comments before I actually check out the code myself.

Since there are still some formatting only changes, I think we should format main first. I would setup some infrastructure for automatic checking/applying of black in a separate PR and afterwards you could rebase on main, dropping all of the formatting related commits entirely w/o reverting anything. How does that sound @anlavandier ?

anlavandier commented:
> I just have some minor comments before I actually check out the code myself.
> 
> Since there are still some formatting only changes, I think we should format main first. I would setup some infrastructure for automatic checking/applying of black in a separate PR and afterwards you could rebase on main, dropping all of the formatting related commits entirely w/o reverting anything. How does that sound @anlavandier ?

That sounds good.

Commit:
Typo in gridToVTK's docstring

renefritze commented:
I've merged the #33 into master just now, please rebase on that @anlavandier when you get the chance.

anlavandier commented:
> I've merged the #33 into master just now, please rebase on that @anlavandier when you get the chance.

I have merged into master into my fork's master. I ran black on my repo and all the files were left unchanged so I didn't have to commit anything.

Commit:
Fixed pointData being used instead of cellData

Commit:
Moved start kwarg at the end

renefritze commented:
I've cleaned up the PR branch and pushed it to https://github.com/pyscience-projects/pyevtk/tree/parallel_vtk 
Would you mind resetting your branch to that @anlavandier ?

anlavandier commented:
Done.

renefritze commented:
> Done.

The PR here still lists the "black" and "revert black" commits. Maybe there was a problem with the update after the reset?

anlavandier commented:
Sorry I misunderstood what you meant by resetting. It should be fine now.

renefritze commented:
> Sorry I misunderstood what you meant by resetting. It should be fine now.

It is. 

Thank you again for the PR. I'll push a new release out shortly.
